### ISSUE TITLE
DRR_WRITE_EMBEDDED records need to include their own byteswap info

### ISSUE THREAD

**Issue by tcaputi** [zfsonlinux/zfs#5854]
### Describe the problem you're observing
Doing a `zfs send -e` from LE to BE and back to LE will cause the dataset to be corrupted. This happens because DRR_WRITE_EMBEDDED records are sent exactly as they are on disk. Both the send and receive code assumes that these embedded blocks will be in the host byteorder. When sending from LE to BE, the BE system knows to store the data with the appropriate byteorder flag set. However, when sending the data back to a LE system, the receive side assumes that all data being sent is in the sending side's byteorder, even though these embedded blocks are not. This also works in reverse (from BE to LE and back).

### Describe how to reproduce the problem
```
[LE system]
zfs create compression=on pool/test
mkdir /pool/test/a
zdb -dddddddd pool/test 7
zfs snap pool/test@now
zfs send -e pool/test@now > sendfile
# move sendfile to other system

[BE system]
zfs recv pool/recv < sendfile
zfs send -e pool/recv@now > sendfile2
# move sendfile2 to other system

[LE system]
zfs recv pool/recv < sendfile2
# at this point you will get warnings about how the dataset can't be mounted
```

This problem has implications for #5769 as well, particularly the raw sends I am working on now. Since encrypted blocks cannot be byteswapped before they are decrypted, raw sends will have to correctly send byteorder information for each block as well (although raw sends will also need to do this for DRR_WRITE and DRR_SPILL blocks as well). My current WIP solution for this is to add a flag `DRR_RAW_BYTESWAP` to `drr_flags` which is set on the send side if `BP_SHOULD_BYTESWAP()`. This can be interpreted on the recv side to detrmine whether or not to set the byteswap flag on the bp.

This solution may not be appropriate for this problem, however, because current systems will not know how to interpret this new flag.

**Comment by tcaputi** [comment]
Technically, it looks like another byteswap bug exists for DRR_WRITE records that are compressed, although its a bit outside normal usage. It seems from looking at `receive_write()` in dmu_send.c that compressed data in the send file can potentially end up getting byteswapped. This shouldn't happen until the data is decompressed. Fortunately, right now the send side will only compress objects that are not metadata, most of which have a byteswap of DMU_BSWAP_UINT8. However, if someone were to send a compressed object of type `DMU_OT_UINT64_OTHER` or `DMU_OTN_UINT*_DATA`, it would fail to send properly.

This code should probably also use a flag similar to the one I describe above, but again this does not fix existing systems.

**Comment by ahrens** [comment]
The best solution to this is to not use `DRR_WRITE_EMBEDDED` for BP's that need to be byteswapped (i.e. are in a different endian from the sending system).  Instead we would decompress them and use a `DRR_WRITE` record.

For the second issue of compressed `DRR_WRITE` records, as you noted that doesn't happen in practice because all non-metadata is UINT8 (except for ztest, which doesn't send/recv).  If it did happen, the receiving code would need to decompress the data before byteswapping it.  To ensure we avoid that in the future we should adjust the code in `do_dump()` to not `request_compressed` if the byteswap func is other than UINT8.

@dankimmel may be interested in this.  Here's the current code that needs to be changed:

```
		/*
		 * We should only request compressed data from the ARC if all
		 * the following are true:
		 *  - stream compression was requested
		 *  - we aren't splitting large blocks into smaller chunks
		 *  - the data won't need to be byteswapped before sending
		 *  - this isn't an embedded block
		 *  - this isn't metadata (if receiving on a different endian
		 *    system it can be byteswapped more easily)
		 */
		boolean_t request_compressed =
		    (dscp->dsc_featureflags & DMU_BACKUP_FEATURE_COMPRESSED) &&
		    !split_large_blocks && !BP_SHOULD_BYTESWAP(bp) &&
		    !BP_IS_EMBEDDED(bp) && !DMU_OT_IS_METADATA(BP_GET_TYPE(bp));
```

**Comment by dankimmel** [comment]
Yeah, I think this patch will fix the issue with `DRR_WRITE` records:
```
diff --git a/usr/src/uts/common/fs/zfs/dmu_send.c b/usr/src/uts/common/fs/zfs/dmu_send.c
index 4da8f636f1..7b378882c0 100644
--- a/usr/src/uts/common/fs/zfs/dmu_send.c
+++ b/usr/src/uts/common/fs/zfs/dmu_send.c
@@ -788,6 +788,7 @@ do_dump(dmu_send_cookie_t *dscp, struct send_range *range)
                 *  - stream compression was requested
                 *  - we aren't splitting large blocks into smaller chunks
                 *  - the data won't need to be byteswapped before sending
+                *  - the data type's byteswap function is DMU_BSWAP_UINT8
                 *  - this isn't an embedded block
                 *  - this isn't metadata (if receiving on a different endian
                 *    system it can be byteswapped more easily)
@@ -795,6 +796,7 @@ do_dump(dmu_send_cookie_t *dscp, struct send_range *range)
                boolean_t request_compressed =
                    (dscp->dsc_featureflags & DMU_BACKUP_FEATURE_COMPRESSED) &&
                    !split_large_blocks && !BP_SHOULD_BYTESWAP(bp) &&
+                   DMU_OT_BYTESWAP(BP_GET_TYPE(bp)) == DMU_BSWAP_UINT8 &&
                    !BP_IS_EMBEDDED(bp) && !DMU_OT_IS_METADATA(BP_GET_TYPE(bp));

                if (BP_IS_EMBEDDED(bp) &&
```
I agree with Matt's approach to the `DRR_WRITE_EMBEDDED` records as well -- not embedding them seems like the simplest approach and shouldn't affect the total amount of data being sent by very much. We could implement that by returning `B_FALSE` from a new case in `send_do_embed()`.

I don't think we have a big endian system to test this on (which is probably how these got broken in the first place :-/). It might be nice to create tests for this that can run on little endian systems (maybe by checking in compressed and embedded send streams from big endian systems? not sure if we have something like this already).

**Comment by ahrens** [comment]
@dankimmel Makes sense.  Could we simplify the `request_compressed` logic to not care about `BP_SHOULD_BYTESWAP` or `DMU_OT_IS_METADATA` and rely on `DMU_OT_BYTESWAP==UINT8` to handle those cases?  Or do we also need to exclude UINT8 metadata?

**Comment by dankimmel** [comment]
Yes, I think that's right. I believe we weren't compressing metadata before basically because it was a superset of the data types that would require byteswap routines `!= UINT8`, and the `BP_SHOULD_BYTESWAP(bp)` check seems unnecessary too.

**Comment by kpande** [comment]
@tcaputi was this addressed?

**Comment by tcaputi** [comment]
No, it hasn't. My rough recollection is that we never decided which solution we actually wanted to implement, but it should be easy to fix once that decision is made.

**Comment by ahrens** [comment]
@pcd1193182 may also want to take a look at this.
